From 10a3079a0fe6e01ae6eb593cac486d545806bdd8 Mon Sep 17 00:00:00 2001
From: Sven Nierlein <sven@nierlein.de>
Date: Fri, 5 Oct 2018 11:51:44 +0200
Subject: [PATCH] fix pending dependencies

right now we did only check the state during dependency checks. But for pending hosts
and services the state is usually OK/UP so the check passed. For pending flag checks
we have to look at the has_been_checked flag as well. This leads to the situation where
services checks will be run if the master service is in pending state even if the service
has pending service execution failure flag set.
---
 src/naemon/checks_host.c        |  10 ++
 src/naemon/checks_service.c     |  12 +++
 4 files changed, 190 insertions(+), 1 deletion(-)
 create mode 100644 tests/test-check-dependencies.c

diff --git a/src/naemon/checks_host.c b/src/naemon/checks_host.c
index d555f092..c0082bc7 100644
--- a/src/naemon/checks_host.c
+++ b/src/naemon/checks_host.c
@@ -232,7 +232,9 @@ static int run_async_host_check(host *hst, int check_options, double latency)
 		}
 
 		/* check host dependencies for execution */
+		log_debug_info(DEBUGL_CHECKS, 0, "Host '%s' checking dependencies...\n", hst->name);
 		if (check_host_dependencies(hst, EXECUTION_DEPENDENCY) == DEPENDENCIES_FAILED) {
+			log_debug_info(DEBUGL_CHECKS, 0, "Host '%s' failed dependency check. Aborting check\n", hst->name);
 			return ERROR;
 		}
 	}
@@ -1157,6 +1159,8 @@ int check_host_dependencies(host *hst, int dependency_type)
 	int state = STATE_UP;
 	time_t current_time = 0L;
 
+	log_debug_info(DEBUGL_CHECKS, 0, "Host '%s' check_host_dependencies()\n", hst->name);
+
 	if (dependency_type == NOTIFICATION_DEPENDENCY) {
 		list = hst->notify_deps;
 	} else {
@@ -1182,10 +1186,16 @@ int check_host_dependencies(host *hst, int dependency_type)
 		else
 			state = temp_host->current_state;
 
+		log_debug_info(DEBUGL_CHECKS, 1, "  depending on host '%s' with state: %d / has_been_checked: %d\n", temp_host->name, state, temp_host->has_been_checked);
+
 		/* is the host we depend on in state that fails the dependency tests? */
 		if (flag_isset(temp_dependency->failure_options, 1 << state))
 			return DEPENDENCIES_FAILED;
 
+		/* check for pending flag */
+		if(temp_host->has_been_checked == FALSE && flag_isset(temp_dependency->failure_options, OPT_PENDING))
+			return DEPENDENCIES_FAILED;
+
 		/* immediate dependencies ok at this point - check parent dependencies if necessary */
 		if (temp_dependency->inherits_parent == TRUE) {
 			if (check_host_dependencies(temp_host, dependency_type) != DEPENDENCIES_OK)
diff --git a/src/naemon/checks_service.c b/src/naemon/checks_service.c
index e1056f1a..93cca540 100644
--- a/src/naemon/checks_service.c
+++ b/src/naemon/checks_service.c
@@ -138,6 +138,8 @@ static void handle_service_check_event(struct nm_event_execution_properties *evp
 	struct timeval event_runtime;
 	int options = temp_service->check_options;
 
+    log_debug_info(DEBUGL_CHECKS, 0, "Service '%s' on host '%s' handle_service_check_event()...\n", temp_service->description, temp_service->host_name);
+
 	if(evprop->execution_type == EVENT_EXEC_NORMAL) {
 
 		/* get event latency */
@@ -189,7 +191,9 @@ static void handle_service_check_event(struct nm_event_execution_properties *evp
 			}
 
 			/* check service dependencies for execution */
+			log_debug_info(DEBUGL_CHECKS, 0, "Service '%s' on host '%s' checking dependencies...\n", temp_service->description, temp_service->host_name);
 			if (check_service_dependencies(temp_service, EXECUTION_DEPENDENCY) == DEPENDENCIES_FAILED) {
+				log_debug_info(DEBUGL_CHECKS, 0, "Service '%s' on host '%s' failed dependency check. Aborting check\n", temp_service->description, temp_service->host_name);
 				return;
 			}
 
@@ -1156,6 +1160,8 @@ int check_service_dependencies(service *svc, int dependency_type)
 	int state = STATE_OK;
 	time_t current_time = 0L;
 
+	log_debug_info(DEBUGL_CHECKS, 0, "Service '%s' on host '%s' check_service_dependencies()\n", svc->description, svc->host_name);
+
 	/* only check dependencies of the desired type */
 	if (dependency_type == NOTIFICATION_DEPENDENCY)
 		list = svc->notify_deps;
@@ -1182,10 +1188,16 @@ int check_service_dependencies(service *svc, int dependency_type)
 		else
 			state = temp_service->current_state;
 
+		log_debug_info(DEBUGL_CHECKS, 1, "  depending on service '%s' on host '%s' with state: %d / has_been_checked: %d\n", temp_service->description, temp_service->host_name, state, temp_service->has_been_checked);
+
 		/* is the service we depend on in state that fails the dependency tests? */
 		if (flag_isset(temp_dependency->failure_options, 1 << state))
 			return DEPENDENCIES_FAILED;
 
+		/* check for pending flag */
+		if(temp_service->has_been_checked == FALSE && flag_isset(temp_dependency->failure_options, OPT_PENDING))
+			return DEPENDENCIES_FAILED;
+
 		/* immediate dependencies ok at this point - check parent dependencies if necessary */
 		if (temp_dependency->inherits_parent == TRUE) {
 			if (check_service_dependencies(temp_service, dependency_type) != DEPENDENCIES_OK)
